src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/RandomNumberGenerator.cs

```diff
@@ -350,6 +353,44 @@ private static void GetHexStringCore(Span<char> destination, bool lowercase)
 
         private static void GetItemsCore<T>(ReadOnlySpan<T> choices, Span<T> destination)
         {
+            // The most expensive part of this operation is the call to get random data. We can
+            // do so only once rather than per element in a common case:
+            // - The number of choices is <= 256. This let's us get a single byte per choice.
+            // - The number of choices is a power of two. This let's us use a byte and simply mask off
+            //   unnecessary bits cheaply rather than needing to use rejection sampling.
+            if (BitOperations.IsPow2(choices.Length) && choices.Length <= 256)
+            {
+                const int StackAllocThreshold = 256;
+
+                // Get a scratch buffer to fill with random bytes, either stack space if the
+                // destination is small enough, or an array pool array if it's larger.
+                // If T is unmanaged, we _could_ use the destination buffer itself, reinterpreted
```

Comment:
`T` is `char` is probably the most interesting case to optimize. I would be in favor of keeping the re-interpret for special casing `char` since random strings are the most interesting case and the impetus for this API.

If we feel that adds too much complexity for a narrow case of "input len is pow2" then that's understandable too.